Patch/block header patch patch (#705)
filecoin-project/specs: src/systems/filecoin_blockchain/struct/block_producer/_index.md

## Block Creation

Scratching (a) winning ElectionPoSt ticket(s), and armed with the requisite `ElectionPoStOutput`, a miner can now publish a new block!

To create a block, the eligible miner must compute a few fields:

- `Parents` - the CIDs of the parent blocks.
- `ParentWeight` - The parent chain's weight (see {{<sref chain_selection>}}).
- `ParentState` - This is the state of the chain after all of the message from each of the `Parents` have been applied to their own `ParentState`. To compute this:
  - Take the `ParentState` of one of the blocks in the chosen parent set (invariant: this is the same value for all blocks in a given parent set).
  - For each block in the parent set, ordered by their tickets:
    - Apply each message in the block to the parent state, in order. If a message was already applied in a previous block, skip it.
    - Transaction fees are given to the miner of the block that the first occurance of the message is included in. If there are two blocks in the parent set, and they both contain the exact same set of messages, the second one will receive no fees.

>     - It is valid for messages in two different blocks of the parent set to conflict, that is, A conflicting message from the combined set of messages will always error.  Regardless of conflicts all messages are applied to the state.

Reviewer: Remove most of this text and link instead to the interpreter's description of tipset execution.

  - For the secpk messages:
    - Insert each of the secpk `SignedMessage`s into a sharray
  - Create a `TxMeta` object and fill each of its fields as follows:
    - `BLSMessages`: the root cid of the bls messages sharray
    - `SECPMessages`: the root cid of the secp messages sharray
  - The cid of this `TxMeta` object should be used to fill the `Messages` field of the block header.
- `BLSAggregate` - The aggregated signatures of all messages in the block that used BLS signing.

> - `Signature` - A signature with the miner's private key (must also match the ticket signature) over the entire block. This is to ensure that nobody tampers with the block after it propagates to the network, since unlike normal PoW blockchains, a winning ticket is found independently of block generation.

Reviewer: miner's worker account private key

Once the miner has the aggregate `ParentState`, they must apply the block reward. This is done by adding the correct block reward amount to the miner owner's account balance in the state tree. The reward will be spendable immediately in this block.

Now, a set of messages is selected to put into the block. For each message, the miner subtracts `msg.GasPrice * msg.GasLimit` from the sender's account balance, returning a fatal processing error if the sender does not have enough funds (this message should not be included in the chain).

Finally, the miner can generate a Unix Timestamp to add to their block, to show that the block generation was appropriately delayed.

Now the block is complete, all that's left is to sign it. The miner serializes the block now (without the signature field), takes the sha256 hash of it, and signs that hash. They place the resultant signature in the `Signature` field.

## Block Broadcast

An eligible miner broadcasts the completed block to the network and assuming everything was done correctly, the network will accept it and other miners will mine on top of it, earning the miner a block reward!

> Miners will need to output their valid block as soon as it is produced, otherwise they risk other miners receiving the block after the EPOCH_CUTOFF and not including them.

Reviewer: ```suggestion
Miners should output their valid block as soon as it is produced, otherwise they risk other miners receiving the block after the EPOCH_CUTOFF and not including them.
```
